operator/standalone/delete.go

```diff
@@ -0,0 +1,98 @@
+package standalone
+
+import (
+	"context"
+	pipeline "github.com/ccremer/go-command-pipeline"
+	helmv1beta1 "github.com/crossplane-contrib/provider-helm/apis/release/v1beta1"
+	"github.com/vshn/appcat-service-postgresql/apis/postgresql/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// DeleteStandalonePipeline is a pipeline that deletes an instance from the target deployment namespace.
+type DeleteStandalonePipeline struct {
+	client             client.Client
+	instance           *v1alpha1.PostgresqlStandalone
+	helmReleaseDeleted *bool
+}
+
+// NewDeleteStandalonePipeline creates a new delete pipeline with the required dependencies.
+func NewDeleteStandalonePipeline(client client.Client, instance *v1alpha1.PostgresqlStandalone) *DeleteStandalonePipeline {
+	return &DeleteStandalonePipeline{
+		instance: instance,
+		client:   client,
+	}
+}
+
+// RunPipeline executes the pipeline with configured business logic steps.
+func (d *DeleteStandalonePipeline) RunPipeline(ctx context.Context) error {
+	return pipeline.NewPipeline().
+		WithSteps(
+			pipeline.NewStepFromFunc("delete connection secret", d.deleteConnectionSecret),
+			pipeline.NewStepFromFunc("delete helm release", d.deleteHelmRelease),
+			pipeline.If(d.isHelmReleaseDeleted,
+				pipeline.NewPipeline().WithNestedSteps("finalize",
+					pipeline.NewStepFromFunc("delete namespace", d.deleteNamespace),
+					pipeline.NewStepFromFunc("remove finalizer", d.removeFinalizer),
+				),
+			),
+		).
+		RunWithContext(ctx).Err()
+}
+
+// deleteHelmRelease removes the Helm Release from the cluster
+//
+func (d *DeleteStandalonePipeline) deleteHelmRelease(ctx context.Context) error {
+	helmRelease := &helmv1beta1.Release{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      d.instance.Status.HelmChart.DeploymentNamespace,
+			Namespace: d.instance.Status.HelmChart.DeploymentNamespace,
+		},
+	}
+	err := d.client.Delete(ctx, helmRelease)
+	if err != nil && apierrors.IsNotFound(err) {
+		d.helmReleaseDeleted = func() *bool { valTrue := true; return &valTrue }()
```

Comment:
Why using a pointer function instead of a plain bool variable? That only leads to more errors since you'd have to check if `d.helmReleaseDeleted` is nil before being able to invoke it...

Reply:
Isn't it not saved if I save a value instead of a pointer?

Reply:
The signature is `func (d *DeleteStandalonePipeline) deleteHelmRelease(ctx context.Context) error`, so `d` is a pointer and thus changes are saved

Reply:
perfect!